Design/histogram_type.md

```diff
@@ -0,0 +1,108 @@
+  - Note that we do not introduce a separate type for histogram and distribution data. This distinction would be only internal, and access to either histogram or distribution data is provided via public interface methods. The goal is to eliminate this distinction from as much of the code as possible.
+- Add a few classes for `X`, `Y`, and `E` (etc.) that would mainly be used internally in the histogram class and for direct access to the underlying data. If these inherit from `std::vetor<double>`, they could also be used to keep the legacy interface for `ISpectrum` alive until the roll-out is complete.
+
+#### Details
+
+- It is probably useful to introduce two types for vectors of standard deviation and variance, instead of the current double use of `E`.
+- Similarly we should have distinct internal types for histogram and distribution data, in case our implementation relies on this distinction. Jon pointed out that it may make most sense to default to distribution data, as OpenGenie did by default.
+- The current copy-on-write (COW) mechanism seems useful and should be preserved.
+  - The `typedef` `MantidVecPtr` for a `cow_ptr` is useless and confusing and should be removed.
+  - We should consider extending the COW mechanism and make sure sharing is preserved more than it is now. For example, what can we do to maintain sharing of `X` for all histograms in a workspace where `X` is modified in an identical way? The COW mechanism as it is now should break down in that case. Does it need to be handled at the workspace level, or can we deal with this here? Having parametric `X` could circumvent that in certain cases (see outlook).
+
+#### Interface
+
+- At least initially not all cases for operations with histograms will be covered by the histogram type or free functions for histograms. Therefore we probably need to provide a direct interface to the underlying data as well.
+- If we want to hide things like error handling, is there a decent way of making this extensible for new operations? Are operators/functions declared as `friend` in our histogram data type? Are they members? Can we allow the use of `std::transform` on, e.g., the `Y` data? How to transform errors?
+
+
+## Roll-out
+
+Since histograms are in widespread use in the Mantid code the introduction of the new data type is not trivial.
+However, we can probably adopt a step-by-step approach that makes the roll-out quite feasible:
+
+- The new histogram data type is typically encapsulated in an `ISpectrum` object.
+  - Currently the `X` data is part of `ISpectrum`, whereas `Y` and `E` are part of `Histogram1D`. We somehow have to change this, or provide the option to have a histogram without `Y` and `E` data. This may be the main difficulty of these first steps.
+  - The interface of `ISpectrum` could be kept unchanged at first. The data access functions would then forward to the internal `X`, `Y`, and `E` data inside the new histogram type.
```

Comment:
I think this is another classic case where `ISpectrum` should be a pure interface and have no implementation of its own. `EventList` & `Histogram1D` can still share some portion of implementation for X if it's in a separate class of its own. Maybe that's a first step?